ls1intum/Artemis
# Feature/adapt structure for multi organizations

### Checklist

### Motivation and Context


From https://github.com/ls1intum/Artemis/issues/2710 

Artemis is currently been used in projects where an instance is hosted for multiple universities and other educational organizations. Some difficulties occur to provide a meaningful separation between them (ex. Students self registration to courses of their own organization, allow instructors to view only users of the same organization, ..).

To simplify these task, keeping an organized structure, a new **organization** entity could be added and mapped to users and courses in order to organize these as sub-groups.

### Description


##### Structure

The proposed solution to this issue is the integration of a new 'Organization' table, where the needed data of multiple organizations can be stored. The organizations are then mapped to users and courses in order to allow the before mentioned separation. 
![UMLClassDiagram (1)](https://user-images.githubusercontent.com/45761921/105817233-9cd73180-5fb5-11eb-8b97-42ddfeeff9e0.png)

##### Usage

Each organization will have an emailPattern column, needed to assign existing , as well as newly registering users, to an organization, without the need of a manually assignment. 

Admins should be able to add organizations to courses, allowing only users/students of the same organizations to register to these.

##### Management

In order to simplify the management of the organizations, a new view will be added for admins to directly add (or remove) users from organizations, as well as creating/updating/removing organizations.

In addition, organizations can be added/removed in the course/user update views directly.

##### Indexing

When artemis is started, the users will be indexed by their email addresses to match currently stored organizations, without the need to manually allocate each user individually to its organization.

Such indexing happens even when the emailPattern of an organization has been updated, updating the users list of the organization.

When a new user is created, it will automatically be assigned to the matching organizations if its email matches the email pattern of the organization.

##### Course registration

The course registration will be carried out in an additive way. In other words, all other course-registration requirements are still required and a matching organization represents an additional requirement on top of the others (if enabled).

### Steps for Testing


1. Log in to Artemis
2. Navigate to Organization management (under Server administration)
3. Test organization managament:
3.1 Create a new organization (users matching the emailPattern should automatically be added)
3.2 Update the organization (users matching the new emailPattern should automatically be added, others not matching removed)
3.3 Remove the organization

4. Click "view" button of an organization to open the detail view
4.1 Try to add a user using the user table
4.2 Try to remove a user using the user table

5. Navigate to course management and edit an active course:
5.1 Add an organization to the course and make sure student registration is enabled
5.2 Save the updated course

6. Navigate to user management and edit a student-user:
6.1 Add the same organization (as previously set in course) to the user
6.2 Save the updated user

7. Using the updated user try to register to the updated course (course-registration.allowed-username-pattern should match too for this step)

### Test Coverage




#### Server

CourseRepository: 90%
OrganizationRepository: 79%
UserRepository: 78%
OrganizationService: 81%
CourseResource: 91%
OrganizationResource: 97%
UserResource: 87%

#### Client

organization-management-detail.component.ts 70%
organization-management-update.component.ts 89%
organization-management.component.ts 96%
organization-management.route.ts 73%
organization-management.service.ts 89%
organization-selector.component.ts 95%





### Screenshots


![Screenshot from 2021-02-24 14-55-21](https://user-images.githubusercontent.com/45761921/109011304-0891e480-76b1-11eb-86f9-1d73cd01c0c3.png)
![Screenshot from 2021-02-24 14-55-35](https://user-images.githubusercontent.com/45761921/109011309-09c31180-76b1-11eb-9445-9feea5707cd2.png)
![Screenshot from 2021-02-24 14-57-58](https://user-images.githubusercontent.com/45761921/109011311-09c31180-76b1-11eb-9f7e-d3ce4b5ef288.png)
![Screenshot from 2021-02-24 14-58-20](https://user-images.githubusercontent.com/45761921/109011313-0a5ba800-76b1-11eb-96f6-e9914842f782.png)
![Screenshot from 2021-02-24 14-58-43](https://user-images.githubusercontent.com/45761921/109011314-0af43e80-76b1-11eb-991d-dde72344dae2.png)
